File: NativeModuleEx/android/app/src/main/java/com/nativemoduleworkshop/CounterManager.kt
package com.nativemoduleworkshop

import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class CounterManager: SimpleViewManager<CounterView>() {
//    React Native 컴포넌트 이름을 반
    override fun getName(): String {
        return REACT_CLASS
    }

//    View를 반환하는 메서드
    override fun createViewInstance(reactContext: ThemedReactContext): CounterView {
//         CounterView.kt 인스턴스를 만들어 반환
        return CounterView(reactContext)
    }

    @ReactProp(name = "leftButtonText")
    fun setLeftButtonText(view: CounterView, text: String){
        view.setLeftButtonText(text)
    }

    @ReactProp(name = "rightButtonText")
    fun setRightButtonText(view: CounterView, text: String){
        view.setRightButtonText(text)
    }

    @ReactProp(name = "value")
    fun setValue(view: CounterView, value: Int){
        view.setValue(value)
    }

    override fun getExportedCustomBubblingEventTypeConstants(): MutableMap<String, Any> {
        val builder = MapBuilder.builder<String, Any>()
        return builder
            .put("pressLeftButton", MapBuilder.of(
                "phasedRegistrationNames",
                MapBuilder.of("bubbled", "onPressLeftButton")
            ))
            .put("pressRightButton", MapBuilder.of(
                "phasedRegistrationNames",
                MapBuilder.of("bubbled", "onPressRightButton")
            )).build()
    }

    companion object{
        const val REACT_CLASS = "Counter"
    }
}

File: NativeModuleEx/ios/CounterManager.m
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(CounterManager, RCTViewManager)
// Props 연동 (연동할 Props 이름, 타입)
  RCT_EXPORT_VIEW_PROPERTY(value, NSNumber)
  RCT_EXPORT_VIEW_PROPERTY(leftButtonText, NSString)
  RCT_EXPORT_VIEW_PROPERTY(rightButtonText, NSString)
// 이벤트를 선언하는 경우 두 번째 파라미터에 RCTDirectEventBlock
  RCT_EXPORT_VIEW_PROPERTY(onPressLeftButton, RCTDirectEventBlock)
  RCT_EXPORT_VIEW_PROPERTY(onPressRightButton, RCTDirectEventBlock)
@end
